File: web/src/filter/conformance.test.ts
/**
 * The conformance fixture, run against this evaluator.
 *
 * `schema/filter-conformance.json` holds a small workspace, a list of filters and the ids
 * each must return. The Go suite inserts the same workspace into Postgres and runs its
 * compiled SQL against the same cases. Neither suite computes the expected answer — it is
 * recorded in the file, so a change that makes both implementations agree on something
 * wrong still fails.
 *
 * Adding a case is the way to fix a filter bug. Fixing one implementation without adding a
 * case leaves the other one wrong, which is the state this file exists to make impossible.
 */

import { readFileSync } from 'node:fs';
import { dirname, join } from 'node:path';
import { fileURLToPath } from 'node:url';

import { describe, expect, it } from 'vitest';

import type { Issue, StateCategory, Timestamp, UUID, WorkflowState } from '~/store/types';

import { filterIssues, type FilterContext } from './evaluate';
import { RELATIVE_KEYWORDS, resolveRelative } from './relative';
import { FILTER_FIELDS, isFilterField } from './types';
import { FilterError, validateFilter } from './validate';

interface FixtureTeam {
  readonly id: UUID;
  readonly key: string;
  readonly name: string;
}

interface FixtureState {
  readonly id: UUID;
  readonly teamId: UUID;
  readonly name: string;
  readonly category: StateCategory;
  readonly position: string;
  readonly isDefault: boolean;
}

interface FixtureIssue {
  readonly id: UUID;
  readonly teamId: UUID;
  readonly number: number;
  readonly title: string;
  readonly description: string;
  readonly stateId: UUID;
  readonly assigneeId?: UUID;
  readonly creatorId?: UUID;
  readonly priority: number;
  readonly estimate?: number;
  readonly dueDate?: string;
  readonly sortOrder: string;
  readonly parentId?: UUID;
  readonly subIssueSortOrder?: string;
  readonly labelIds: readonly UUID[];
  readonly templateId?: UUID;
  readonly recurringIssueId?: UUID;
  readonly completedAt?: Timestamp;
  readonly archivedAt?: Timestamp;
  readonly deletedAt?: Timestamp;
  readonly createdAt: Timestamp;
  readonly updatedAt: Timestamp;
}

interface FixtureRelation {
  readonly issueId: UUID;
  readonly relatedIssueId: UUID;
  readonly type: string;
}

interface FixtureSubscription {
  readonly issueId: UUID;
  readonly userId: UUID;
  readonly unsubscribed: boolean;
}

interface FixtureCustomer {
  readonly id: UUID;
  readonly status: string;
  readonly tier?: string;
  readonly revenue?: number;
  readonly size?: number;
}

interface FixtureCustomerRequest {
  readonly issueId: UUID;
  readonly customerId?: UUID;
  readonly important: boolean;
}

interface Fixture {
  readonly evaluatedAt: Timestamp;
  readonly timezone: string;
  /** Token to the exact UTC instant it resolves to at `evaluatedAt`. `$comment` is lines. */
  readonly relativeTokens: Readonly<Record<string, string | readonly string[]>>;
  readonly workspace: { readonly id: UUID };
  readonly teams: readonly FixtureTeam[];
  readonly workflowStates: readonly FixtureState[];
  readonly issues: readonly FixtureIssue[];
  readonly relations: readonly FixtureRelation[];
  readonly subscriptions: readonly FixtureSubscription[];
  readonly customers: readonly FixtureCustomer[];
  readonly customerRequests: readonly FixtureCustomerRequest[];
  readonly cases: readonly {
    readonly name: string;
    readonly filter: unknown;
    readonly expect: readonly string[];
  }[];
  readonly errors: readonly {
    readonly name: string;
    readonly filter: unknown;
    readonly message: string;
  }[];
}

/**
 * Read rather than imported: the fixture is the schema's, shared with the Go suite, and
 * `import`ing it would inline a copy into the bundle graph and give TypeScript a literal
 * type for a file whose shape is declared above anyway.
 */
// Resolved from this file rather than from the working directory so the suite runs the
// same whether vitest was started in `web/` or at the repository root. Deliberately not
// `new URL(path, import.meta.url)`: vite rewrites that form into an asset URL served over
// http, which node:fs cannot open.
const FIXTURE_PATH = join(
  dirname(fileURLToPath(import.meta.url)),
  '../../../schema/filter-conformance.json',
);

const fixture = JSON.parse(readFileSync(FIXTURE_PATH, 'utf8')) as Fixture;

/**
 * Short names expand to `01900000-0000-7000-8000-` + the name left-padded to twelve.
 *
 * Written out in full these arrays are unreadable, and an unreadable fixture is one nobody
 * checks when it disagrees with them. Both loaders expand the same way.
 */
const ID_PREFIX = '01900000-0000-7000-8000-';
const SHORT_NAME = /^[0-9a-f]{1,12}$/;

function expandId(name: string): UUID {
  return SHORT_NAME.test(name) ? ID_PREFIX + name.padStart(12, '0') : name;
}

/**
 * Expands short names inside a filter's `values`, and only for the fields that hold ids.
 *
 * Type-driven rather than shape-driven on purpose: `"1"` is a short name by every pattern
 * you could write, and expanding a priority into a uuid would make the fixture's simplest
 * case fail for a reason nobody would find.
 */
function expandFilter(node: unknown): unknown {
  if (typeof node !== 'object' || node === null) return node;
  const record = node as Record<string, unknown>;

  const nodes = record['nodes'];
  if (Array.isArray(nodes)) return { ...record, nodes: nodes.map(expandFilter) };

  const field = record['field'];
  const values = record['values'];
  if (
    typeof field === 'string' &&
    isFilterField(field) &&
    FILTER_FIELDS[field].type === 'uuid' &&
    Array.isArray(values)
  ) {
    return {
      ...record,
      values: values.map((value: unknown) => (typeof value === 'string' ? expandId(value) : value)),
    };
  }
  return record;
}

const NOW: Timestamp = '2026-01-01T00:00:00Z';

const teamKeys = new Map<UUID, string>(fixture.teams.map((team) => [team.id, team.key]));

const issues: Issue[] = fixture.issues.map((row) => ({
  id: row.id,
  workspaceId: fixture.workspace.id,
  teamId: row.teamId,
  number: row.number,
  identifier: `${teamKeys.get(row.teamId) ?? '?'}-${row.number}`,
  title: row.title,
  description: row.description,
  stateId: row.stateId,
  assigneeId: row.assigneeId,
  creatorId: row.creatorId,
  priority: row.priority,
  sortOrder: row.sortOrder,
  estimate: row.estimate,
  dueDate: row.dueDate,
  dueDateSource: 'manual',
  parentId: row.parentId,
  subIssueSortOrder: row.subIssueSortOrder,
  templateId: row.templateId,
  recurringIssueId: row.recurringIssueId,
  completedAt: row.completedAt,
  archivedAt: row.archivedAt,
  createdAt: row.createdAt,
  updatedAt: row.updatedAt,
}));

const states = new Map<UUID, WorkflowState>(
  fixture.workflowStates.map((row) => [
    row.id,
    {
      id: row.id,
      workspaceId: fixture.workspace.id,
      teamId: row.teamId,
      name: row.name,
      color: '#888888',
      category: row.category,
      position: row.position,
      isDefault: row.isDefault,
      isSystem: false,
      createdAt: NOW,
      updatedAt: NOW,
    },
  ]),
);

function collect(rows: Iterable<readonly [UUID, UUID]>): ReadonlyMap<UUID, ReadonlySet<UUID>> {
  const out = new Map<UUID, Set<UUID>>();
  for (const [key, value] of rows) {
    const bucket = out.get(key);
    if (bucket === undefined) out.set(key, new Set([value]));
    else bucket.add(value);
  }
  return out;
}

function addToSet(map: Map<UUID, Set<string>>, issueId: UUID, value: string): void {
  const bucket = map.get(issueId);
  if (bucket === undefined) map.set(issueId, new Set([value]));
  else bucket.add(value);
}

const byCustomer = new Map(fixture.customers.map((row) => [row.id, row]));

const customerMaps = (() => {
  const customers = new Map<UUID, Set<UUID>>();
  const counts = new Map<UUID, number>();
  const statuses = new Map<UUID, Set<string>>();
  const tiers = new Map<UUID, Set<string>>();
  const revenues = new Map<UUID, number[]>();
  const sizes = new Map<UUID, number[]>();
  const important = new Set<UUID>();
  for (const request of fixture.customerRequests ?? []) {
    counts.set(request.issueId, (counts.get(request.issueId) ?? 0) + 1);
    if (request.important) important.add(request.issueId);
    if (request.customerId === undefined) continue;
    const customer = byCustomer.get(request.customerId);
    if (customer === undefined) continue;
    addToSet(customers, request.issueId, customer.id);
    addToSet(statuses, request.issueId, customer.status);
    if (customer.tier !== undefined) addToSet(tiers, request.issueId, customer.tier);
    if (customer.revenue !== undefined) {
      const bucket = revenues.get(request.issueId);
      if (bucket === undefined) revenues.set(request.issueId, [customer.revenue]);
      else bucket.push(customer.revenue);
    }
    if (customer.size !== undefined) {
      const bucket = sizes.get(request.issueId);
      if (bucket === undefined) sizes.set(request.issueId, [customer.size]);
      else bucket.push(customer.size);
    }
  }
  return { customers, counts, statuses, tiers, revenues, sizes, important };
})();

const context: FilterContext = {
  time: { now: Date.parse(fixture.evaluatedAt), timezone: fixture.timezone },
  states,
  labels: collect(fixture.issues.flatMap((row) => row.labelIds.map((id) => [row.id, id] as const))),
  // An explicit unsubscribe is a flag rather than a missing row, so the flagged rows are
  // dropped here: `subscriber eq grace` must not return an issue she unsubscribed from.
  subscribers: collect(
    fixture.subscriptions
      .filter((row) => !row.unsubscribed)
      .map((row) => [row.issueId, row.userId] as const),
  ),
  // Only `blocks` is stored, and "blocked by" is the same row read from the other end.
  blockedBy: collect(
    fixture.relations
      .filter((row) => row.type === 'blocks')
      .map((row) => [row.relatedIssueId, row.issueId] as const),
  ),
  blocking: collect(
    fixture.relations
      .filter((row) => row.type === 'blocks')
      .map((row) => [row.issueId, row.relatedIssueId] as const),
  ),
  customers: customerMaps.customers,
  customerCount: customerMaps.counts,
  customerStatus: customerMaps.statuses,
  customerTier: customerMaps.tiers,
  customerRevenue: customerMaps.revenues,
  customerSize: customerMaps.sizes,
  customerImportant: customerMaps.important,
  // The replica normally holds no deleted issues at all; the fixture supplies them the way
  // a trash view would, so that `deleted eq true` has something to find.
  deleted: new Set(
    fixture.issues.filter((row) => row.deletedAt !== undefined).map((row) => row.id),
  ),
};

describe('filter conformance', () => {
  it('loaded the fixture the Go suite runs', () => {
    expect(fixture.cases.length).toBeGreaterThan(0);
    expect(fixture.errors.length).toBeGreaterThan(0);
    expect(issues).toHaveLength(fixture.issues.length);
  });

  for (const testCase of fixture.cases) {
    it(testCase.name, () => {
      const filter = validateFilter(expandFilter(testCase.filter));
      // Compared as a set, sorted: ordering is a display option and is tested separately.
      // Mixing the two here would make an ordering change look like a filter regression.
      const got = filterIssues(issues, filter, context).sort();
      expect(got).toEqual(testCase.expect.map(expandId).sort());
    });
  }
});

describe('filter conformance errors', () => {
  for (const testCase of fixture.errors) {
    it(testCase.name, () => {
      // Rejected, not silently ignored. An ignored clause widens the result set, and a
      // filter that quietly matches more than it says is the bug that makes people stop
      // trusting filters.
      let thrown: unknown;
      try {
        validateFilter(expandFilter(testCase.filter));
      } catch (error) {
        thrown = error;
      }
      expect(thrown, 'the filter must be rejected').toBeInstanceOf(FilterError);
      // The fixture pins a fragment of the message so the two implementations cannot drift
      // into rejecting the same input for different stated reasons.
      expect((thrown as FilterError).message).toContain(testCase.message);
    });
  }
});

/**
 * Every relative token, resolved, against the instants recorded in the fixture.
 *
 * The cases above already compare id sets, and this asks a stricter question of a smaller
 * thing: not "do the two evaluators select the same issues" but "do they resolve the same
 * token to the same instant". The fixture holds seven issues, and with seven issues a great
 * many wrong instants still select the right rows — a resolver that ignored the workspace
 * timezone would pass every date case above, because nothing in the fixture was created
 * within an hour of midnight.
 *
 * It also reaches what the case list structurally cannot. A case using a token only one side
 * accepts fails on the side that rejects it, so nobody writes one, so exactly the divergence
 * this file exists to prevent is the one it cannot see. Five tokens sat in that blind spot:
 * this evaluator shipped `now`, `yesterday`, `tomorrow`, `startOfMonth` and `startOfYear`,
 * and the server's `parseRelative` refused all five — so the filter bar built filters that
 * worked here and that `CreateView` then declined to save.
 *
 * services/internal/filter/relative_conformance_test.go reads the same table.
 */
describe('relative tokens resolve to the instants the fixture records', () => {
  const recorded = Object.entries(fixture.relativeTokens).filter(
    (entry): entry is [string, string] => typeof entry[1] === 'string',
  );

  it('covers every keyword this evaluator accepts', () => {
    // Otherwise a token can be added to both implementations and pinned by neither, which
    // is the state that produced the divergence in the first place.
    const pinned = new Set(recorded.map(([token]) => token));
    for (const keyword of RELATIVE_KEYWORDS) {
      expect(pinned.has(keyword), `no instant recorded for "${keyword}"`).toBe(true);
    }
  });

  for (const [token, expected] of recorded) {
    it(`${token} resolves to ${expected}`, () => {
      const { instant } = resolveRelative(token, context.time);
      expect(new Date(instant).toISOString()).toBe(new Date(expected).toISOString());
    });
  }
});

File: services/internal/filter/conformance_test.go
package filter_test

import (
	"context"
	"encoding/json"
	"os"
	"path/filepath"
	"slices"
	"strings"
	"testing"
	"time"

	// The fixture pins a timezone by name. A machine without a zoneinfo database would
	// fall back to UTC without saying so and move every relative date by an hour, so the
	// table is embedded and the test answers the same question everywhere it runs.
	_ "time/tzdata"

	"github.com/google/uuid"
	"github.com/jackc/pgx/v5"
	"github.com/jackc/pgx/v5/pgxpool"

	"github.com/peixotolabs/polaris/services/internal/filter"
	"github.com/peixotolabs/polaris/services/internal/testutil"
)

// The conformance fixture, run against a real Postgres.
//
// schema/filter-conformance.json is the contract between this package and web/src/filter.
// Neither suite computes the expected answer: it is recorded in the file, so a change that
// makes both implementations agree on something wrong still fails. That is the whole point
// of the file — two evaluators that agree with each other and disagree with the spec are
// exactly as broken as one that is simply wrong, and much harder to notice.
//
// The workspace goes in through raw SQL rather than through the domain layer. The fixture
// names its own ids and carries archived and soft-deleted issues, neither of which the
// create path produces, and routing it through the domain layer would make this test fail
// whenever issue creation had a bug — a conformance failure pointing at the wrong
// subsystem.

// idPrefix expands the fixture's short names: "e1" is ...-0000000000e1 and "a11" is
// ...-000000000a11. Written out in full those arrays are unreadable, and an unreadable
// fixture is one nobody checks when it disagrees with them.
const idPrefix = "01900000-0000-7000-8000-"

// uuidFields are the fields whose values are ids, and therefore the ones whose values get
// expanded.
//
// Decided by field rather than by the shape of the string. A pattern match would be a
// guess, and the day somebody adds {"field":"title","op":"contains","values":["e2"]} the
// guess turns a text search into a uuid and the case fails for a reason nobody can see.
var uuidFields = map[string]bool{
	"state": true, "assignee": true, "creator": true, "subscriber": true,
	"label": true, "team": true, "parent": true, "blockedBy": true, "blocking": true,
}

type conformance struct {
	EvaluatedAt time.Time `json:"evaluatedAt"`
	Timezone    string    `json:"timezone"`

	Workspace struct {
		ID     uuid.UUID `json:"id"`
		Name   string    `json:"name"`
		URLKey string    `json:"urlKey"`
		Plan   string    `json:"plan"`
	} `json:"workspace"`

	Users []struct {
		ID   uuid.UUID `json:"id"`
		Name string    `json:"name"`
		Role string    `json:"role"`
	} `json:"users"`

	Teams []struct {
		ID   uuid.UUID `json:"id"`
		Key  string    `json:"key"`
		Name string    `json:"name"`
	} `json:"teams"`

	WorkflowStates []struct {
		ID        uuid.UUID `json:"id"`
		TeamID    uuid.UUID `json:"teamId"`
		Name      string    `json:"name"`
		Category  string    `json:"category"`
		Position  string    `json:"position"`
		IsDefault bool      `json:"isDefault"`
	} `json:"workflowStates"`

	Labels []struct {
		ID       uuid.UUID  `json:"id"`
		Name     string     `json:"name"`
		Color    string     `json:"color"`
		Position string     `json:"position"`
		IsGroup  bool       `json:"isGroup"`
		ParentID *uuid.UUID `json:"parentId"`
		TeamID   *uuid.UUID `json:"teamId"`
	} `json:"labels"`

	Issues []struct {
		ID                uuid.UUID   `json:"id"`
		TeamID            uuid.UUID   `json:"teamId"`
		Number            int64       `json:"number"`
		Title             string      `json:"title"`
		Description       string      `json:"description"`
		StateID           uuid.UUID   `json:"stateId"`
		AssigneeID        *uuid.UUID  `json:"assigneeId"`
		CreatorID         *uuid.UUID  `json:"creatorId"`
		Priority          int         `json:"priority"`
		Estimate          *int        `json:"estimate"`
		DueDate           *string     `json:"dueDate"`
		ParentID          *uuid.UUID  `json:"parentId"`
		SortOrder         string      `json:"sortOrder"`
		SubIssueSortOrder *string     `json:"subIssueSortOrder"`
		LabelIDs          []uuid.UUID `json:"labelIds"`
		CompletedAt       *time.Time  `json:"completedAt"`
		ArchivedAt        *time.Time  `json:"archivedAt"`
		DeletedAt         *time.Time  `json:"deletedAt"`
		CreatedAt         time.Time   `json:"createdAt"`
		UpdatedAt         time.Time   `json:"updatedAt"`
	} `json:"issues"`

	Relations []struct {
		ID             uuid.UUID `json:"id"`
		IssueID        uuid.UUID `json:"issueId"`
		RelatedIssueID uuid.UUID `json:"relatedIssueId"`
		Type           string    `json:"type"`
	} `json:"relations"`

	Subscriptions []struct {
		ID           uuid.UUID `json:"id"`
		IssueID      uuid.UUID `json:"issueId"`
		UserID       uuid.UUID `json:"userId"`
		Reason       string    `json:"reason"`
		Unsubscribed bool      `json:"unsubscribed"`
	} `json:"subscriptions"`

	Cases []struct {
		Name   string          `json:"name"`
		Filter json.RawMessage `json:"filter"`
		Expect []string        `json:"expect"`
	} `json:"cases"`

	Errors []struct {
		Name    string          `json:"name"`
		Filter  json.RawMessage `json:"filter"`
		Message string          `json:"message"`
	} `json:"errors"`
}

func loadConformance(t *testing.T) *conformance {
	t.Helper()
	path := filepath.Join("..", "..", "..", "schema", "filter-conformance.json")
	raw, err := os.ReadFile(path)
	if err != nil {
		t.Fatalf("read %s: %v", path, err)
	}
	var c conformance
	if err := json.Unmarshal(raw, &c); err != nil {
		t.Fatalf("parse %s: %v", path, err)
	}
	return &c
}

// expandID turns a short fixture name into the id it stands for, and leaves anything
// already written out in full alone.
func expandID(name string) string {
	if len(name) == len(idPrefix)+12 {
		return name
	}
	return idPrefix + strings.Repeat("0", 12-len(name)) + name
}

// expandFilter rewrites the short ids inside a case's filter so the compiler sees the same
// bytes a real client would send.
func expandFilter(t *testing.T, raw json.RawMessage) json.RawMessage {
	t.Helper()

	var node map[string]any
	if err := json.Unmarshal(raw, &node); err != nil {
		t.Fatalf("parse case filter: %v", err)
	}
	expandNode(t, node)

	out, err := json.Marshal(node)
	if err != nil {
		t.Fatalf("re-encode case filter: %v", err)
	}
	return out
}

func expandNode(t *testing.T, node map[string]any) {
	t.Helper()

	if field, ok := node["field"].(string); ok && uuidFields[field] {
		values, _ := node["values"].([]any)
		for i, v := range values {
			s, ok := v.(string)
			if !ok {
				t.Fatalf("case filter: %s value %d is not a string", field, i)
			}
			values[i] = expandID(s)
		}
	}
	children, _ := node["nodes"].([]any)
	for _, child := range children {
		m, ok := child.(map[string]any)
		if !ok {
			t.Fatalf("case filter: a group's node is not an object")
		}
		expandNode(t, m)
	}
}

func TestConformance(t *testing.T) {
	t.Parallel()

	fx := loadConformance(t)
	loc, err := time.LoadLocation(fx.Timezone)
	if err != nil {
		t.Fatalf("load timezone %s: %v", fx.Timezone, err)
	}

	db := testutil.NewDB(t)
	ctx := context.Background()
	insertWorkspace(t, ctx, db.Pool(), fx)

	for _, tc := range fx.Cases {
		t.Run(tc.Name, func(t *testing.T) {
			node, err := filter.Parse(expandFilter(t, tc.Filter))
			if err != nil {
				t.Fatalf("parse: %v", err)
			}

			compiled, err := filter.Compile(node, filter.Options{
				Now:      fx.EvaluatedAt,
				Location: loc,
				// The workspace scope belongs to the caller, exactly as it does in the
				// domain layer: this package compiles a fragment, never a whole query.
				ArgOffset: 1,
			})
			if err != nil {
				t.Fatalf("compile: %v", err)
			}

			query := `SELECT id FROM issue WHERE workspace_id = $1 AND ` + compiled.SQL
			args := append([]any{fx.Workspace.ID}, compiled.Args...)

			rows, err := db.Pool().Query(ctx, query, args...)
			if err != nil {
				t.Fatalf("query: %v\n%s\nargs: %v", err, query, args)
			}
			got, err := pgx.CollectRows(rows, pgx.RowTo[uuid.UUID])
			if err != nil {
				t.Fatalf("scan: %v\n%s", err, query)
			}

			gotIDs := make([]string, 0, len(got))
			for _, id := range got {
				gotIDs = append(gotIDs, id.String())
			}
			wantIDs := make([]string, 0, len(tc.Expect))
			for _, name := range tc.Expect {
				wantIDs = append(wantIDs, expandID(name))
			}
			slices.Sort(gotIDs)
			slices.Sort(wantIDs)

			// Compared as a set. Ordering is a display option and is tested separately;
			// mixing the two here would make an ordering change look like a filter
			// regression.
			if !slices.Equal(gotIDs, wantIDs) {
				t.Fatalf("id set mismatch\n got: %v\nwant: %v\nsql:  %s\nargs: %v",
					shortIDs(gotIDs), shortIDs(wantIDs), compiled.SQL, compiled.Args)
			}
		})
	}

	for _, tc := range fx.Errors {
		t.Run("rejects/"+tc.Name, func(t *testing.T) {
			_, err := filter.Parse(expandFilter(t, tc.Filter))
			if err == nil {
				t.Fatalf("accepted a filter that must be rejected: %s", tc.Filter)
			}
			// A rejection carrying an unrecognisable message is only marginally better
			// than no rejection: whoever wrote the filter has to be told which part of it
			// the server could not read.
			if !strings.Contains(err.Error(), tc.Message) {
				t.Fatalf("message %q does not mention %q", err.Error(), tc.Message)
			}
		})
	}
}

// A needle is a substring, not a pattern.
//
// The fixture cannot pin this, because both evaluators would have to agree on what a
// wildcard means and neither offers one. What it pins instead is that Postgres honours the
// escaping the compiler emits: filter_test.go asserts the escaped needle is bound, and
// this asserts the database then reads it literally. Without both halves, somebody
// searching for "50%" gets every issue whose title starts with 50.
func TestContainsIsASubstringNotAPattern(t *testing.T) {
	t.Parallel()

	fx := loadConformance(t)
	db := testutil.NewDB(t)
	ctx := context.Background()
	insertWorkspace(t, ctx, db.Pool(), fx)

	literal := fx.Issues[0].ID
	decoy := fx.Issues[1].ID
	if _, err := db.Pool().Exec(ctx,
		`UPDATE issue SET title = CASE id WHEN $1 THEN 'Fix the 50% _redirect' ELSE 'Fix the 50XX yredirect' END
		 WHERE id IN ($1, $2)`, literal, decoy); err != nil {
		t.Fatalf("retitle: %v", err)
	}

	node, err := filter.Parse([]byte(`{"field":"title","op":"contains","values":["50% _redirect"]}`))
	if err != nil {
		t.Fatalf("parse: %v", err)
	}
	compiled, err := filter.Compile(node, filter.Options{ArgOffset: 1})
	if err != nil {
		t.Fatalf("compile: %v", err)
	}

	rows, err := db.Pool().Query(ctx,
		`SELECT id FROM issue WHERE workspace_id = $1 AND `+compiled.SQL,
		append([]any{fx.Workspace.ID}, compiled.Args...)...)
	if err != nil {
		t.Fatalf("query: %v", err)
	}
	got, err := pgx.CollectRows(rows, pgx.RowTo[uuid.UUID])
	if err != nil {
		t.Fatalf("scan: %v", err)
	}
	if len(got) != 1 || got[0] != literal {
		t.Fatalf("the wildcards were honoured rather than escaped: got %v, want [%v]", got, literal)
	}
}

// shortIDs trims the shared prefix back off, so a failure reads in the fixture's own
// vocabulary instead of as seven near-identical uuids.
func shortIDs(ids []string) []string {
	out := make([]string, 0, len(ids))
	for _, id := range ids {
		out = append(out, strings.TrimLeft(strings.TrimPrefix(id, idPrefix), "0"))
	}
	return out
}

// insertWorkspace writes the fixture with raw SQL, in dependency order.
func insertWorkspace(t *testing.T, ctx context.Context, pool *pgxpool.Pool, fx *conformance) {
	t.Helper()

	exec := func(sql string, args ...any) {
		t.Helper()
		if _, err := pool.Exec(ctx, sql, args...); err != nil {
			t.Fatalf("fixture insert: %v\n%s\nargs: %v", err, sql, args)
		}
	}

	exec(`INSERT INTO workspace (id, name, url_key, plan) VALUES ($1, $2, $3, $4)`,
		fx.Workspace.ID, fx.Workspace.Name, fx.Workspace.URLKey, fx.Workspace.Plan)

	for _, u := range fx.Users {
		exec(`INSERT INTO "user" (id, workspace_id, name, display_name, role) VALUES ($1, $2, $3, $3, $4)`,
			u.ID, fx.Workspace.ID, u.Name, u.Role)
	}

	for _, tm := range fx.Teams {
		exec(`INSERT INTO team (id, workspace_id, key, name) VALUES ($1, $2, $3, $4)`,
			tm.ID, fx.Workspace.ID, tm.Key, tm.Name)
	}

	for _, s := range fx.WorkflowStates {
		exec(`INSERT INTO workflow_state (id, workspace_id, team_id, name, category, position, is_default)
		      VALUES ($1, $2, $3, $4, $5, $6, $7)`,
			s.ID, fx.Workspace.ID, s.TeamID, s.Name, s.Category, s.Position, s.IsDefault)
	}

	// Groups before their children: label_parent_integrity reads the parent row and
	// refuses a parent that is absent or is not a group. The fixture lists them in that
	// order and this loop preserves it.
	for _, l := range fx.Labels {
		exec(`INSERT INTO label (id, workspace_id, team_id, parent_id, is_group, name, color, position)
		      VALUES ($1, $2, $3, $4, $5, $6, $7, $8)`,
			l.ID, fx.Workspace.ID, l.TeamID, l.ParentID, l.IsGroup, l.Name, l.Color, l.Position)
	}

	// Parents before children, for the same reason: issue_parent_acyclic walks the chain.
	for _, is := range fx.Issues {
		var dueDate *time.Time
		if is.DueDate != nil {
			d, err := time.Parse("2006-01-02", *is.DueDate)
			if err != nil {
				t.Fatalf("fixture issue %s: due date %q: %v", is.ID, *is.DueDate, err)
			}
			dueDate = &d
		}

		exec(`INSERT INTO issue (
		        id, workspace_id, team_id, number, title, description, state_id,
		        assignee_id, creator_id, priority, estimate, due_date, parent_id,
		        sort_order, sub_issue_sort_order,
		        completed_at, archived_at, deleted_at, created_at, updated_at)
		      VALUES ($1, $2, $3, $4, $5, $6, $7, $8, $9, $10, $11, $12, $13, $14, $15,
		              $16, $17, $18, $19, $20)`,
			is.ID, fx.Workspace.ID, is.TeamID, is.Number, is.Title, is.Description, is.StateID,
			is.AssigneeID, is.CreatorID, is.Priority, is.Estimate, dueDate, is.ParentID,
			is.SortOrder, is.SubIssueSortOrder,
			is.CompletedAt, is.ArchivedAt, is.DeletedAt, is.CreatedAt, is.UpdatedAt)

		// team_id and group_id are left to issue_label_denormalise, which is the only
		// writer that can be trusted to keep them in step with the label tree.
		for _, labelID := range is.LabelIDs {
			exec(`INSERT INTO issue_label (id, workspace_id, issue_id, label_id) VALUES ($1, $2, $3, $4)`,
				uuid.Must(uuid.NewV7()), fx.Workspace.ID, is.ID, labelID)
		}
	}

	for _, r := range fx.Relations {
		exec(`INSERT INTO issue_relation (id, workspace_id, issue_id, related_issue_id, type)
		      VALUES ($1, $2, $3, $4, $5)`,
			r.ID, fx.Workspace.ID, r.IssueID, r.RelatedIssueID, r.Type)
	}

	for _, s := range fx.Subscriptions {
		exec(`INSERT INTO issue_subscription (id, workspace_id, issue_id, user_id, reason, unsubscribed)
		      VALUES ($1, $2, $3, $4, $5, $6)`,
			s.ID, fx.Workspace.ID, s.IssueID, s.UserID, s.Reason, s.Unsubscribed)
	}
}
